contrib/platform/src/com/sun/jna/platform/win32/WinInetUtil.java

```diff
@@ -0,0 +1,104 @@
+/* Copyright (c) 2015 Michael Freeman, All Rights Reserved
+ * 
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * 
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.  
+ */
+package com.sun.jna.platform.win32;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.sun.jna.Native;
+import com.sun.jna.platform.win32.WinInet.INTERNET_CACHE_ENTRY_INFO;
+import com.sun.jna.platform.win32.WinNT.HANDLE;
+import com.sun.jna.ptr.IntByReference;
+
+/**
+ * Reusable functions that use WinInet.dll
+ * 
+ * @author mlfreeman[at]gmail.com
+ */
+public final class WinInetUtil {
+	/**
+	 * Helper function for traversing wininet's cache and returning all entries.
+	 * <br>
+	 * Some entries are cookies, some entries are history items, and some are
+	 * actual files.<br>
+	 * 
+	 * @return A map of cache URL => local file (or URL => empty string for
+	 *         cookie and history entries)
+	 */
+	public static Map<String, String> getBrowserCache() {
+		List<INTERNET_CACHE_ENTRY_INFO> items = new ArrayList<WinInet.INTERNET_CACHE_ENTRY_INFO>();
+
+		// WinInet APIs could return these two errors:
+
+		// this one happens if the INTERNET_CACHE_ENTRY_INFO structure will
+		// require more space to be filled properly
+		// ERROR_INSUFFICIENT_BUFFER
+		// 122 (0x7A)
+		// The data area passed to a system call is too small.
+
+		// this one happens when you've gone through the cache completely
+		// ERROR_NO_MORE_ITEMS
+		// 259 (0x103)
+		// No more data is available.
+
+		IntByReference size = new IntByReference();
+		HANDLE entryHandle = WinInet.INSTANCE.FindFirstUrlCacheEntry(null, null, size);
```

Comment:
According to the documentation https://msdn.microsoft.com/en-us/library/windows/desktop/aa384026(v=vs.85).aspx:

> Returns a handle that the application can use in the FindNextUrlCacheEntry function to retrieve subsequent entries in the cache. If the function fails, the return value is NULL. To get extended error information, call GetLastError..

Your code should check for this and handle the error - especially ERROR_INSUFFICIENT_BUFFER - e.g.:
```java
HANDLE h = FindFirstUrlCacheEntry(...);
if (h == NULL) {
    int hr = GetLastError();
    if (hr == ERROR_INSUFFICIENT_BUFFER ) {
        ...re-allocate and try again (?)
    } else {
        throw new Win32Exception(...failed: hr);
    }
}
```